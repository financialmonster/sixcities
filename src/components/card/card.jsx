import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import {Link} from 'react-router-dom';
import cn from 'classnames';
import {bindActionCreators} from 'redux';

import {focusCard, blurCard} from '../../actions/helpers.js';

const Card = ({offer: {price, is_premium: isPremium, is_favorite: isFavorite, rating, title, type,
    preview_image: src, id}, cardMouseEnterHandler, cardMouseLeaveHandler}, isNearby) => {
    const btnClassName = cn(`place-card__bookmark-button button`, {'place-card__bookmark-button--active': isFavorite});
    const articleClassName = cn(`place-card`, {'near-places__card': isNearby, 'cities__place-card': !isNearby});
    const divClassName = cn(`place-card__image-wrapper`,
        {'near-places__image-wrapper': isNearby, 'cities__image-wrapper': !isNearby});

    return (
        <Link to={`/offer/${id}`} title={`To the ${title} offer page`}>
            <article className={articleClassName} onMouseEnter={() => cardMouseEnterHandler(id)}
                onMouseLeave={cardMouseLeaveHandler}
            >
                {(isPremium) ? <div className="place-card__mark"><span>Premium</span></div> : null}

                <div className={divClassName}>
                    <img className="place-card__image" src={src} width="260" height="200" alt="Place pic" />
                </div>
                <div className="place-card__info">
                    <div className="place-card__price-wrapper">
                        <div className="place-card__price">
                            <b className="place-card__price-value">&euro;{price}</b>
                            <span className="place-card__price-text">&#47;&nbsp;night</span>
                        </div>
                        <button className={btnClassName} type="button"
                            title={(isFavorite) ? `In bookmarks` : `To bookmarks`}
                        >
                            <svg className="place-card__bookmark-icon" width="18" height="19">
                                <use xlinkHref="#icon-bookmark"></use>
                            </svg>
                            <span className="visually-hidden">{(isFavorite) ? `In bookmarks` : `Not in bookmarks`}</span>
                        </button>
                    </div>
                    <div className="place-card__rating rating">
                        <div className="place-card__stars rating__stars">
                            <span style={{width: `${20 * rating}%`}}></span>
                            <span className="visually-hidden">Rating</span>
                        </div>
                    </div>
                    <h2 className="place-card__name">{title}</h2>
                    <p className="place-card__type">{type}</p>
                </div>
            </article>
        </Link>
    )
}

Card.propTypes = {
    cardMouseEnterHandler: PropTypes.func.isRequired,
    cardMouseLeaveHandler: PropTypes.func.isRequired,
    isNearby: PropTypes.bool.isRequired,
    offer: PropTypes.shape({
        price: PropTypes.number.isRequired,
        is_favorite: PropTypes.bool.isRequired,
        is_premium: PropTypes.bool.isRequired,
        rating: PropTypes.number.isRequired,
        title: PropTypes.string.isRequired,
        type: PropTypes.string.isRequired,
        preview_image: PropTypes.string.isRequired,
        id: PropTypes.number.isRequired
    })
}

const mapDispatchToProps = (dispatch) => {
    return {
        cardMouseEnterHandler:  bindActionCreators(focusCard ,dispatch),
        cardMouseLeaveHandler:  bindActionCreators(blurCard, dispatch)
    }
}

export {Card};
export default connect(null, mapDispatchToProps)(Card);
