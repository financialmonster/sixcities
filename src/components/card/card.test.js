import React from 'react';
import renderer from 'react-test-renderer';
import {BrowserRouter as Router} from 'react-router-dom';

import {Card} from './card.jsx';

const offers = [
    {
        price: 120,
        is_premium: true,
        is_favorite: false,
        rating: 4,
        title: `Beautiful luxurious apartment at great location`,
        type: `Apartment`,
        preview_image: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg`,
        src: `img/apartment-01.jpg`,
        id: 0
    },
    {
        price: 80,
        is_premium: false,
        is_favorite: false,
        rating: 4,
        title: `Wood and stone place`,
        type: `Private room`,
        preview_image: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg`,
        src: `img/room.jpg`,
        id: 1
    },
    {
        price: 132,
        is_premium: false,
        is_favorite: true,
        rating: 4,
        title: `Canal View Prinsengracht`,
        type: `Apartment`,
        preview_image: `https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg`,
        src: `img/apartment-02.jpg`,
        id: 2
    }
]

describe(`Correctly renders Card component which is located on the offerPage`, () => {
    it(`which is neither favorite nor premium`, () => {
        const tree = renderer.create(
            <Router>
                <Card offer={offers[1]} cardMouseEnterHandler={() => {}}
                    cardMouseLeaveHandler={() => {}} isMain={true} isNearby={false} isFavPage={false}
                />
            </Router>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`which is not favorite but premium`, () => {
        const tree = renderer.create(
            <Router>
                <Card offer={offers[0]} cardMouseEnterHandler={() => {}}
                    cardMouseLeaveHandler={() => {}} isMain={true} isNearby={false} isFavPage={false}
                />
            </Router>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });   
    
    it(`which is not premium but favorite`, () => {
        const tree = renderer.create(
            <Router>
                <Card offer={offers[2]} cardMouseEnterHandler={() => {}}
                    cardMouseLeaveHandler={() => {}} isMain={true} isNearby={false} isFavPage={false}
                />
            </Router>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });
});

describe(`Correctly renders Card component which is located on the mainPage`, () => {
    it(`which is neither favorite nor premium`, () => {
        const tree = renderer.create(
            <Router>
                <Card offer={offers[1]} cardMouseEnterHandler={() => {}}
                    cardMouseLeaveHandler={() => {}} isMain={false} isNearby={true} isFavPage={false}
                />
            </Router>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`which is not favorite but premium`, () => {
        const tree = renderer.create(
            <Router>
                <Card offer={offers[0]} cardMouseEnterHandler={() => {}}
                    cardMouseLeaveHandler={() => {}} isMain={false} isNearby={true} isFavPage={false}
                />
            </Router>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });   
    
    it(`which is not premium but favorite`, () => {
        const tree = renderer.create(
            <Router>
                <Card offer={offers[2]} cardMouseEnterHandler={() => {}}
                    cardMouseLeaveHandler={() => {}} isMain={false} isNearby={true} isFavPage={false}
                />
            </Router>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });
});

describe(`Correctly renders Card component which is located on the favoritesPage`, () => {
    it(`which is neither favorite nor premium`, () => {
        const tree = renderer.create(
            <Router>
                <Card offer={offers[1]} cardMouseEnterHandler={() => {}}
                    cardMouseLeaveHandler={() => {}} isMain={false} isNearby={false} isFavPage={true}
                />
            </Router>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });

    it(`which is not favorite but premium`, () => {
        const tree = renderer.create(
            <Router>
                <Card offer={offers[0]} cardMouseEnterHandler={() => {}}
                    cardMouseLeaveHandler={() => {}} isMain={false} isNearby={false} isFavPage={true}
                />
            </Router>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });   
    
    it(`which is not premium but favorite`, () => {
        const tree = renderer.create(
            <Router>
                <Card offer={offers[2]} cardMouseEnterHandler={() => {}}
                    cardMouseLeaveHandler={() => {}} isMain={false} isNearby={false} isFavPage={true}
                />
            </Router>)
        .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
