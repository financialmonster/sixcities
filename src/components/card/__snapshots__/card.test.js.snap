// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly renders Card component which is located on the favoritesPage which is neither favorite nor premium 1`] = `
<a
  href="/offer/1"
  onClick={[Function]}
  title="To the Wood and stone place offer page"
>
  <article
    className="place-card favorites__card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__image-wrapper favorites__image-wrapper"
    >
      <img
        alt="Place pic"
        className="place-card__image"
        height="110"
        src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg"
        width="150"
      />
    </div>
    <div
      className="place-card__info favorites__card-info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            80
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          title="To bookmarks"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            Not in bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "80%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        Wood and stone place
      </h2>
      <p
        className="place-card__type"
      >
        Private room
      </p>
    </div>
  </article>
</a>
`;

exports[`Correctly renders Card component which is located on the favoritesPage which is not favorite but premium 1`] = `
<a
  href="/offer/0"
  onClick={[Function]}
  title="To the Beautiful luxurious apartment at great location offer page"
>
  <article
    className="place-card favorites__card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__mark"
    >
      <span>
        Premium
      </span>
    </div>
    <div
      className="place-card__image-wrapper favorites__image-wrapper"
    >
      <img
        alt="Place pic"
        className="place-card__image"
        height="110"
        src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg"
        width="150"
      />
    </div>
    <div
      className="place-card__info favorites__card-info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            120
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          title="To bookmarks"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            Not in bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "80%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        Beautiful luxurious apartment at great location
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
</a>
`;

exports[`Correctly renders Card component which is located on the favoritesPage which is not premium but favorite 1`] = `
<a
  href="/offer/2"
  onClick={[Function]}
  title="To the Canal View Prinsengracht offer page"
>
  <article
    className="place-card favorites__card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__image-wrapper favorites__image-wrapper"
    >
      <img
        alt="Place pic"
        className="place-card__image"
        height="110"
        src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg"
        width="150"
      />
    </div>
    <div
      className="place-card__info favorites__card-info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            132
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button place-card__bookmark-button--active"
          title="In bookmarks"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            In bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "80%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        Canal View Prinsengracht
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
</a>
`;

exports[`Correctly renders Card component which is located on the mainPage which is neither favorite nor premium 1`] = `
<a
  href="/offer/1"
  onClick={[Function]}
  title="To the Wood and stone place offer page"
>
  <article
    className="place-card near-places__card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__image-wrapper near-places__image-wrapper"
    >
      <img
        alt="Place pic"
        className="place-card__image"
        height="200"
        src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg"
        width="260"
      />
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            80
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          title="To bookmarks"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            Not in bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "80%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        Wood and stone place
      </h2>
      <p
        className="place-card__type"
      >
        Private room
      </p>
    </div>
  </article>
</a>
`;

exports[`Correctly renders Card component which is located on the mainPage which is not favorite but premium 1`] = `
<a
  href="/offer/0"
  onClick={[Function]}
  title="To the Beautiful luxurious apartment at great location offer page"
>
  <article
    className="place-card near-places__card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__mark"
    >
      <span>
        Premium
      </span>
    </div>
    <div
      className="place-card__image-wrapper near-places__image-wrapper"
    >
      <img
        alt="Place pic"
        className="place-card__image"
        height="200"
        src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg"
        width="260"
      />
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            120
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          title="To bookmarks"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            Not in bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "80%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        Beautiful luxurious apartment at great location
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
</a>
`;

exports[`Correctly renders Card component which is located on the mainPage which is not premium but favorite 1`] = `
<a
  href="/offer/2"
  onClick={[Function]}
  title="To the Canal View Prinsengracht offer page"
>
  <article
    className="place-card near-places__card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__image-wrapper near-places__image-wrapper"
    >
      <img
        alt="Place pic"
        className="place-card__image"
        height="200"
        src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg"
        width="260"
      />
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            132
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button place-card__bookmark-button--active"
          title="In bookmarks"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            In bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "80%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        Canal View Prinsengracht
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
</a>
`;

exports[`Correctly renders Card component which is located on the offerPage which is neither favorite nor premium 1`] = `
<a
  href="/offer/1"
  onClick={[Function]}
  title="To the Wood and stone place offer page"
>
  <article
    className="place-card cities__place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__image-wrapper cities__image-wrapper"
    >
      <img
        alt="Place pic"
        className="place-card__image"
        height="200"
        src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/3.jpg"
        width="260"
      />
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            80
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          title="To bookmarks"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            Not in bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "80%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        Wood and stone place
      </h2>
      <p
        className="place-card__type"
      >
        Private room
      </p>
    </div>
  </article>
</a>
`;

exports[`Correctly renders Card component which is located on the offerPage which is not favorite but premium 1`] = `
<a
  href="/offer/0"
  onClick={[Function]}
  title="To the Beautiful luxurious apartment at great location offer page"
>
  <article
    className="place-card cities__place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__mark"
    >
      <span>
        Premium
      </span>
    </div>
    <div
      className="place-card__image-wrapper cities__image-wrapper"
    >
      <img
        alt="Place pic"
        className="place-card__image"
        height="200"
        src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/9.jpg"
        width="260"
      />
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            120
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button"
          title="To bookmarks"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            Not in bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "80%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        Beautiful luxurious apartment at great location
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
</a>
`;

exports[`Correctly renders Card component which is located on the offerPage which is not premium but favorite 1`] = `
<a
  href="/offer/2"
  onClick={[Function]}
  title="To the Canal View Prinsengracht offer page"
>
  <article
    className="place-card cities__place-card"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    <div
      className="place-card__image-wrapper cities__image-wrapper"
    >
      <img
        alt="Place pic"
        className="place-card__image"
        height="200"
        src="https://htmlacademy-react-3.appspot.com/six-cities/static/hotel/4.jpg"
        width="260"
      />
    </div>
    <div
      className="place-card__info"
    >
      <div
        className="place-card__price-wrapper"
      >
        <div
          className="place-card__price"
        >
          <b
            className="place-card__price-value"
          >
            €
            132
          </b>
          <span
            className="place-card__price-text"
          >
            / night
          </span>
        </div>
        <button
          className="place-card__bookmark-button button place-card__bookmark-button--active"
          title="In bookmarks"
          type="button"
        >
          <svg
            className="place-card__bookmark-icon"
            height="19"
            width="18"
          >
            <use
              xlinkHref="#icon-bookmark"
            />
          </svg>
          <span
            className="visually-hidden"
          >
            In bookmarks
          </span>
        </button>
      </div>
      <div
        className="place-card__rating rating"
      >
        <div
          className="place-card__stars rating__stars"
        >
          <span
            style={
              Object {
                "width": "80%",
              }
            }
          />
          <span
            className="visually-hidden"
          >
            Rating
          </span>
        </div>
      </div>
      <h2
        className="place-card__name"
      >
        Canal View Prinsengracht
      </h2>
      <p
        className="place-card__type"
      >
        Apartment
      </p>
    </div>
  </article>
</a>
`;
